### PR TITLE
Use classnames package for dynamic className props

### PR DESCRIPTION
**Pull request by Shavindra**
## Summary

Addresses issue #1072. See #1105 for extra context/previous PR comments.

## Relevant technical choices

Refactored to use `classname` npm library.

## Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

**Comment by Shavindra** [comment]
@googlebot I fixed it

File: assets/js/components/button.js [hunk 1]
```diff
@@ -60,12 +60,15 @@ class Button extends Component {
 
 		return (
 			<SemanticButton
-				className={ `
-					mdc-button
-					${ ! text && 'mdc-button--raised' }
-					${ className && className }
-					${ danger ? 'mdc-button--danger' : '' }
-				` }
+
```
Comment (tofumatt) [inline assets/js/components/button.js]: This empty line can be removed.

File: assets/js/components/dashboard-splash/dashboard-splash-service.js [hunk 2]
```diff
@@ -36,15 +37,21 @@ class DashboardSplashService extends Component {
...
-							mdc-layout-grid__cell
-							mdc-layout-grid__cell--span-5-tablet
-							mdc-layout-grid__cell--span-7-desktop
-							${ opposite ? 'mdc-layout-grid__cell--start-3-tablet mdc-layout-grid__cell--start-5-desktop mdc-layout-grid__cell--offset-5-tablet' : '' }
-						` }>
+						<div className={ classnames(
+							'mdc-layout-grid__cell',
+							'mdc-layout-grid__cell--span-5-tablet',
+							'mdc-layout-grid__cell--span-7-desktop',
+							{ 'mdc-layout-grid__cell--start-3-tablet mdc-layout-grid__cell--start-5-desktop mdc-layout-grid__cell--offset-5-tablet': opposite }
```
Comment (tofumatt) [inline assets/js/components/dashboard-splash/dashboard-splash-service.js]: As mentioned in https://github.com/google/site-kit-wp/pull/1105/files#r377545433:

It would be cleaner to read this with each `className` as its own line. I know it's a bit "redundant" but it'll great improve readability, eg:

```suggestion
							{
								'mdc-layout-grid__cell--start-3-tablet': opposite.
								'mdc-layout-grid__cell--start-5-desktop': opposite,
								'mdc-layout-grid__cell--offset-5-tablet': opposite,
							}
```

File: assets/js/components/data-block.js [hunk 3]
```diff
@@ -71,15 +72,14 @@ class DataBlock extends Component {
...
 			<div
-				className={ `
-					googlesitekit-data-block
-					googlesitekit-data-block--${ context }
-					${ selected ? 'googlesitekit-data-block--selected' : '' }
-					${ className }
-				` }
+				className={ classnames(
+					'googlesitekit-data-block',
+					className,
```
Comment (tofumatt) [inline assets/js/components/data-block.js]: You previously mentioned that changing the order of these class names messed with the CSS specificity for some of the classes (see: https://github.com/google/site-kit-wp/pull/1105/files#r377040444).

Was that not the case after all or is this an oversight? I thought the order here needed to be preserved—if that isn't the case: great! But could you elaborate on what the issue was/what you thought it was and whether this is an intentional/okay change to make after all? 😄

File: assets/js/components/data-block.js [hunk 4]
```diff
@@ -71,15 +72,14 @@ class DataBlock extends Component {
 		} = this.props;
 
 		const role = ( 'button' === context ) ? 'button' : '';
-
```
Comment (tofumatt) [inline assets/js/components/data-block.js]: Any reason for removing this? It seems unrelated and I've always liked a newline separating out the `render()` logic from its return value/`render()` output.

I'd say this should be restored.

File: assets/js/components/notifications/cta.js [hunk 5]
```diff
@@ -39,10 +40,12 @@ class CTA extends Component {
 		} = this.props;
 
 		return (
-			<div className={ `
```
Comment (tofumatt) [inline assets/js/components/notifications/cta.js]: Very nice getting rid of all of these multiline template string `className` values. Thanks! 👍

File: assets/js/components/data-block.js (hunk 3 above)
Comment (Shavindra) [reply assets/js/components/data-block.js]: When I initially did that it broke and I had layouts getting messed up. @aaemnnosttv pointed out it shouldn't happen. I think it's also because tests were failing after I reordered initially I thought it was specificity. I was wrong. doh!

Comment (tofumatt) [reply assets/js/components/data-block.js]: Ah, awesome! Good to know, as it's always a pain redoing weird specificity rules. Sounds good and like this is sorted, thanks for the explanation 👍

**Comment by tofumatt** [comment]
To clarify on the recommended usage of `classnames` in the files I mentioned: one of the benefits of using `classnames` is greater readability of many class names being applied to a component. Many of the files I pointed out have static `className`s (or are just using template string interpolation), but they have 5+ separate class names. Take https://github.com/Shavindra/site-kit-wp/blob/9cb4448a811ab58880b99bd47447e9fba5dec6e9/assets/js/components/page-header.js#L75-L82, for instance. It'd be easier to read:

```js
<div className={ classnames(
    'mdc-layout-grid__cell',
    'mdc-layout-grid__cell--align-bottom',
    'mdc-layout-grid__cell--align-right-tablet',
    'mdc-layout-grid__cell--span-4-phone',
    'mdc-layout-grid__cell--span-4-tablet',
    'mdc-layout-grid__cell--span-6-desktop',
) }>
```

compared to:

```js
<div className={ `
    mdc-layout-grid__cell
    mdc-layout-grid__cell--align-bottom
    mdc-layout-grid__cell--align-right-tablet
    mdc-layout-grid__cell--span-4-phone
    mdc-layout-grid__cell--span-4-tablet
    mdc-layout-grid__cell--span-6-desktop
` }>
```

It's also easier to work with the `classnames` version. Finally, it makes our usage of anything other than a single, static `className` consistent throughout the codebase. Since there aren't many left to change and it'll give a greater sense of consistency to the code, I think it's worth doing 😄

**Comment by Shavindra** [comment]
> To clarify on the recommended usage of `classnames` in the files I mentioned: one of the benefits 

I think it would be an overkill to wrap simple strings with `classnames` when you can just simply have the classes on each line anyway? Also we don't need the template strings e.g. 

https://github.com/google/site-kit-wp/blob/9aef263e74be80c9023f32fbdf8d9644a73737d5/assets/js/components/settings/settings-admin.js#L65

https://github.com/google/site-kit-wp/blob/9aef263e74be80c9023f32fbdf8d9644a73737d5/assets/js/components/header.js#L42

As you can see in this file for e.g. they are just normal strings.

**Comment by tofumatt** [comment]
Fair point! There are a lot of those in non-template strings it looks like. I think a follow-up issue to clean them up is worthwhile to file, could you please do that instead of making those multi-line string changes in this PR @Shavindra?

Feel free to skip the multiline strings here, but there are a few other little points mentioned in my review to address. After that this should be good to merge.

File: assets/js/components/data-block.js (hunk 4 above)
Comment (Shavindra) [reply assets/js/components/data-block.js]: This is removed already. But it seems github is not picking it up properly.

File: assets/js/components/data-table.js [hunk 6]
```diff
@@ -91,13 +92,17 @@ export const getDataTableFromData = ( data, headers, options ) => {
...
-				` }
-			>
+		<div className={ classnames(
+			'googlesitekit-table',
+			{ 'googlesitekit-table--with-list': ! options || ! options.disableListMode }
+		) }>
+			<table className={ classnames(
+				'googlesitekit-table__wrapper',
+				{
+					'googlesitekit-table__wrapper-1': ! data && ! data[ 0 ].length,
```
Comment (tofumatt) [inline assets/js/components/data-table.js]: This logic doesn't follow the old one and can't even be true. If `data` is falsey it can't also have an array accessor method. Even an empty array in JS is truthy, so `! data && data[ 0 ]` will always be false… and if `data` is `null` or `undefined` this will throw a `TypeError`. The aim here is to show a single column when the `data` used to populate the table is false-y (`null`/`undefined`) or has no elements inside it (`! data[ 0 ].length`). It's also missing the check to make sure `data[ 0 ]` exists.

This conditional should be: `! data || ! data[ 0 ] || ! data[ 0 ].length`.

File: assets/js/components/data-table.js [hunk 7]
```diff
@@ -91,13 +92,17 @@ export const getDataTableFromData = ( data, headers, options ) => {
...
+					[ `googlesitekit-table__wrapper--${ data[ 0 ].length }-col` ]: data && data[ 0 ].length,
```
Comment (tofumatt) [inline assets/js/components/data-table.js]: This should check for `data[ 0 ]` as well:
```suggestion
					[ `googlesitekit-table__wrapper--${ data[ 0 ].length }-col` ]: data && data[ 0 ] && data[ 0 ].length,
```

File: assets/js/components/data-table.js (hunk 6 above)
Comment (Shavindra) [reply assets/js/components/data-table.js]: yes you're right. I copied and pasted.

**Comment by aaemnnosttv** [comment]
@googlebot I consent.

File: assets/js/components/dialog.js [hunk 8]
```diff
@@ -66,7 +67,10 @@ class Dialog extends Component {
 		return (
 			<div
 				ref={ this.dialogRef }
-				className={ `mdc-dialog ${ dialogActive ? 'mdc-dialog--open' : '' }` }
+				className={ classnames(
+					'mdc-dialog',
+					{ 'mdc-dialog--open': dialogActive } )
```
Comment (tofumatt) [inline assets/js/components/dialog.js]: I can't find the original comment but saw a PR email about this one…

It feels like something ESLint _should_ complain about and certainly can, but I'm guessing we don't have that sort of rule enabled.

I'd happily add it, but I don't feel it's worth blocking the PR on 😄

File: assets/js/components/data-table.js [hunk 9]
```diff
@@ -91,13 +92,17 @@ export const getDataTableFromData = ( data, headers, options ) => {
...
-			>
+		<div className={ classnames(
+			'googlesitekit-table',
+			{ 'googlesitekit-table--with-list': ! options || ! options.disableListMode }
+		) }>
+			<table className={ classnames(
+				'googlesitekit-table__wrapper',
+				{
+					'googlesitekit-table__wrapper--1-col': ! data || ! data[ 0 ] || ! data[ 0 ].length,
+					[ `googlesitekit-table__wrapper--${ data[ 0 ].length }-col` ]: data && data[ 0 ] && data[ 0 ].length,
```
Comment (felixarntz) [inline assets/js/components/data-table.js]: Couldn't this be problematic? I'm not entirely sure, but I would think this object property gets evaluated regardless, which would be problematic if e.g. `! data`.

Comment (tofumatt) [reply assets/js/components/data-table.js]: Huh, that's a good catch! (I really wish this had a test for `false`-y `data` values. But that's true; it will still get evaluated and `TypeError`.

I'd say it's best to refactor this component to have a `column` value that's used here instead, eg:

```js
const columns = data && data[ 0 ] && data[ 0 ].length ? data[ 0 ].length : 1;

[...]
<table className={ classnames(
  'googlesitekit-table__wrapper',
  `googlesitekit-table__wrapper--${ columns }-col`
) }>
```

Good catch 👍

**Comment by Shavindra** [comment]
@googlebot I fixed it.

File: assets/js/components/data-table.js [hunk 10]
```diff
@@ -90,14 +91,17 @@ export const getDataTableFromData = ( data, headers, options ) => {
 		);
 	} );
 
+	const columns = data && data[ 0 ] && data[ 0 ].length ? data[ 0 ].length : 1;
```
Comment (tofumatt) [inline assets/js/components/data-table.js]: Thanks; that's nicer to read anyway 😄

**Comment by tofumatt** [comment]
@googlebot I consent.
